DOCS-7815 applied template (#238)
mulesoft/docs-api-manager: modules/ROOT/pages/message-logging-policy.adoc

include::_attributes.adoc[]
endif::[]

[width="100%", cols="5,15"]
|===

> >s| Policy Name | Message Logging

Reviewer: Table column headers use title-style capitals but row leadins use sentence style:

```suggestion
>s| Policy name | Message Logging
```

Author: Done! Thanks!

>s|Summary      | Logs custom messages using the information from incoming requests
>s|Category | Troubleshooting
>s|First Mule version available | 3.8.0
.1+>.^s| Returned Status Codes
|None
|===

> This policy logs a custom message with the information available from policies, proxy, and the backend, at any point of the execution. This policy works with both Apache log4j and Apache log4j2.

Reviewer: the proxy? proxies?

Author: The Message Logging policy logs custom messages using the information from incoming requests, the response from the backend, or the information from other policies applied to the same API endpoint.

== Prerequisites

> You must have the Anypoint Platform organization administrator (OrgAdmin) role or have permissions to create or manage APIs.

Reviewer: ```suggestion
You must have the Anypoint Platform organization administrator role (OrgAdmin) or have permission to create or manage APIs.
```

Author: Done! Thanks!

== How This Policy Works

Logging a policy consists of the following sequence:

. A request is sent to the API.

> . The message is logged, if the following conditions are true:

Reviewer: ```suggestion
. The message is logged if the following conditions are true:
```

Author: Done! Thanks!

** The level and category of the log defined in the policy are included in the logger defined in the log4j2.xml configuration file.

> ** The conditional field of the policy was not set, or was set, and the condition was met.

Reviewer: ```suggestion
** Either the conditional field of the policy was not set, or it was set and the condition was met.
```

Author: Done! Thanks!

. The API response is returned. Message Logging Policy does not interfere with the execution of any policy nor its flow.
. The message appears in the application log.

== About the Logging Format

A message log uses the following format: 
+
`<date> <thread name> <level> <category>: <message>`
+

> `date`: The date and timestamp when the message is logged.

Reviewer: ```suggestion
`date` is the message logging date and timestamp.
```

Author: Done! Thanks!

> `thread name`: The event thread for which the message is logged.

Reviewer: ```suggestion
`thread name` is the name of the event thread in which the message appears.
```

Author: Done! Thanks!

> `level`: The logging level or type of the message, which is the same as log4j levels.

Reviewer: ```suggestion
`level` is the logging level or type of the message (log4j level).
```

Author: Done! Thanks!

+
For more information, see https://www.tutorialspoint.com/log4j/log4j_logging_levels.htm[log4j Logging Levels].
//Nahuel--what are the various levels?

> `category`: This parameter value is always `org.mule`.

Reviewer 2: From the docs: "This parameter is optional. If no category is set, the default is the org.mule.runtime.logging.policy-<policy id> value."

Author: Done! Thanks!

> `message`: This parameter value is always "the payload".

Reviewer 2: It is not a fixed value. "the payload" is just an example of a fixed message.

Author: Done! Thanks!

> `message`: This parameter value is always "the payload".

Reviewer: `message`is always "the payload".

But...isn't your tech reviewer saying something else?

Author: Done! Thanks!

include::_attributes.adoc[]
endif::[]

[width="100%", cols="5,15"]
|===
>s| Policy Name | Message Logging

> >s|Summary      | Logs messages user-password defined on LDAP

Reviewer 2: "Logs custom messages. Content can be composed using information from the incoming request."
... Or similar.

Author: Done! Thanks!

> >s|Category | Security

Reviewer 2: Category is `Troubleshooting`.

Author: Done! Thanks!

> >s|First Mule version available | 4.0.0

Reviewer 2: Since 3.8.0. (originally feature was for Mule 4 and then it was backported to Mule 3)

Reviewer 2: Not sure if this is to be left as is or change it to reflect that is also available in Mule 3.

Author: Done! Thanks!

> >s|Summary      | Logs custom messages using information from incoming requests, the backend, or information from other policies applied to the same API endpoint.

Reviewer: ```suggestion
>s|Summary      | Logs custom messages using information from incoming requests, responses from the back end, or information from other policies applied to the same API endpoint.
```

Author: No, backend is one word.

.1+>.^s| Returned Status Codes
|401 - Unauthorized or invalid client application credentials

> // what is the right ret code for this policy

Reviewer 2: Since it only logs info, the policy does not change the status code at all.

Author: Done! Thanks!

> The Message Logging policy logs custom messages using information from incoming requests, the response from the backend, or information from other policies applied to the same API endpoint. 

Reviewer: ```suggestion
The Message Logging policy logs custom messages using information from incoming requests, responses from the back end, or information from other policies applied to the same API endpoint. 
```

Author: Done! Thanks!

This policy logs a custom message with the information available between policies, proxy, or backend at any point of the execution. This policy works with both Apache log4j and Apache log4j2.

> //Nahuel--what is meant by information available between policies, proxy, or backend at any point of the execution

Reviewer 2: Proxy communicates with a backend (unless the backend itself is the one to be tracked). Policies can be applied to a proxy in a layered (ordered) way. This is a policy so it can be  applied at certain point among policies, by using the "order" feature. So, this policy has the information available from the incoming request and policies currently applied before it, and/or information coming from the backend, and any policies currently applied between it  and the message policy.

Author: Done! Thanks!

== Prerequisites

You must have the Anypoint Platform organization administrator (OrgAdmin) role or have permissions to create or manage APIs.

== How This Policy Works

Logging a policy consists of the following sequence:

. A request is sent to the API.
. The message is logged, if the following conditions are true:
** The level and category of the log defined in the policy are included in the logger defined in the log4j2.xml configuration file.
** The conditional field of the policy was not set, or was set, and the condition was met.
. The API response is returned. Message Logging Policy does not interfere with the execution of any policy nor its flow.
. The message appears in the application log.

== About the Logging Format

A message log uses the following format: 
+
`<date> <thread name> <level> <category>: <message>`
+
`date`: The date and timestamp when the message is logged.
`thread name`: The event thread for which the message is logged.
`level`: The logging level or type of the message.

> //Nahuel--what are the various levels?

Reviewer 2: Levels are the ones defined for log4j. For example: https://www.tutorialspoint.com/log4j/log4j_logging_levels.htm

Author: Done! Thanks!

> `category`: This parameter value that typically indicates the package and class where the logging occurred. The default value is `org.mule`.

Reviewer: ```suggestion
`category` typically indicates the package and class where the logging occurred. The default value is `org.mule`.
```

Author: Done! Thanks!

> `message`: This parameter displays the details of the type of issue, for example, "the payload".

Reviewer: ```suggestion
`message` displays the details of the type of issue, for example, `the payload`:
```

Author: Done! Thanks!

+

> Example:

Reviewer: ```suggestion

```

Author: Done! Thanks!

`2018-01-25 18:57:29,907 [WrapperListener_start_runner] INFO org.mule: the payload`

> == Understanding the Logging States

Reviewer: ```suggestion
== Logging States
```
...but consider something more precise: "When Message Logging Occurs" or "Events That Cause Message Logging" or similar.

Author: Logging States During Events

> The message logging occurs at various events for a request. These logging states include:

Reviewer: ```suggestion
Message logging occurs at specific intervals during a request flow:
```

Reviewer: Should the following be list items?

Author: I would leave it as is. Done! Thanks!

== Prerequisites

You must have the Anypoint Platform Organization Administrators role or have permission to create or manage APIs.

== How This Policy Works

Logging a policy consists of the following sequence:

. A request is sent to the API.
. The message is logged if the following conditions are true:
** The level and category of the log defined in the policy are included in the logger defined in the log4j2.xml configuration file.
** Either the conditional field of the policy was not set, or was set and the condition was met.
. The API response is returned. Message Logging Policy does not interfere with the execution of any policy nor its flow.
. The message appears in the application log.

== About the Logging Format

A message log uses the following format: 
+
`<date> <thread name> <level> <category>: <message>`
+
`date` is the message logging date and timestamp.
`thread name` is the name of the event thread in which the message appears.
`level` is the logging level or type of the message, such as INFO, WARN, ERROR, and DEBUG (log4j level).
+
For more information, see https://www.tutorialspoint.com/log4j/log4j_logging_levels.htm[log4j Logging Levels].

> `category` typically indicates the package and class where the logging occurred. The default value is `org.mule`.

Reviewer 2: The default value is not `org.mule` but `org.mule.runtime.logging.policy-<policy id>` as stated in this doc later.

Author: Done! Thanks!

> `message`: displays the details of the type of issue, for example, `the payload`:

Reviewer 2: It's not necessarily an issue or type of issue. Can be anything you want to log at a certain point.

Author: Done! Thanks!

image::apim-message-logging-policy-states-3.png[State after the error handler executes]

== Configuring Policy Parameters

When you apply the Message Logging policy to your API from the UI, you can configure the following parameters:

image::apim-message-logging-policy-ui.png[Message Logging Policy UI]

[%header%autowidth.spread]
|===
|Parameter |Description
|Message |A DataWeave expression that's transformed into a message to be logged.
Note: If the payload is used as part of the message log, do not configure the Listener as non-repeatable (by default this setting is off).
|Conditional |The condition expressed as a DataWeave expression that returns true or false that needs to be satisfied to log the message. This field is optional.
|Category |The category that the log has. This parameter is optional. If no category is set, the default is the `org.mule.runtime.logging.policy-<policy id>` value.

> |Level |Defines if the message is logged according to the configuration in your log4j2 file.

Reviewer 2: The levels values are the same as the ones used by log4j, but they are not in the log4j file. Both levels and file are two different things in this case.

Author: Done! Thanks!

> |Before Calling API |Set to cause your logging policy to be applied before the API is called.

Reviewer 2: Partially true statement. I would say something like the following, instead: "Set to cause your logging policy to log messages before the API endpoint is called, taking into account the order in which this policy is applied".

Author: Done! Thanks!

> |After Calling API |Set to cause your logging policy to be applied after the API is called.

Reviewer 2: Partially true statement. I would say something like the following, instead: "Set to cause your logging policy to log messages after the API endpoint is called, taking into account the order in which this policy is applied"

Author: Done! Thanks!

image::apim-message-logging-policy-states-3.png[State after the error handler executes]

== Configuring Policy Parameters

> When you apply the Message Logging policy to your API from the UI, you can configure the following parameters:

Reviewer: ```suggestion
When you use the UI to apply the Message Logging policy to your API, you can configure the following parameters:
```

Author: Done! Thanks!

image::apim-message-logging-policy-ui.png[Message Logging Policy UI]

[%header%autowidth.spread]
|===

> |Parameter |Description | Required

Reviewer: ```suggestion
|Parameter |Description | Required?
```

Author: Done! Thanks!

> |Message |A DataWeave expression to use for extracting the message to be logged.

Reviewer: ```suggestion
|Message |A DataWeave expression to use for extracting the message to be logged
```

Author: Done! Thanks!

> Note: If the payload is used as part of the message log, do not configure the Listener as non-repeatable (by default this setting is off).|Required

Reviewer: ```suggestion
Note: If the payload is used as part of the message log, do not configure the Listener as nonrepeatable (by default this setting is off).|Required
```
Do we need to say that this setting is off, esp. if telling them to NOT set it?

Author: Done! Thanks!

> |Conditional |A DataWeave Boolean expression used to specify whether a message is to be logged. |Optional.

Reviewer: No periods at the end of fragment descriptions
```suggestion
|Conditional |A DataWeave Boolean expression used to specify whether a message is to be logged |Optional
```

Author: Done! Thanks!

> |Category |A prefix value in the log message that indicates the package and class where the log is executed. If no category is set, the default value is `org.mule.runtime.logging.policy-<policy id>`. |Optional

Reviewer: ```suggestion
|Category |A prefix in the log message that indicates the package and class where the log is executed
+ 
+ If no category is set, the default value is `org.mule.runtime.logging.policy-<policy id>`. |Optional
```

Author: Done! Thanks!

> |Level |Defines the level of message to log, INFO, WARN, ERROR, or DEBUG. |Required

Reviewer: ```suggestion
|Level |Defines the level of message to log: INFO, WARN, ERROR, or DEBUG |Required
```

Author: Done! Thanks!

> |Before Calling API |Set to cause your logging policy to log messages before the API endpoint is called, taking into account the order in which this policy is applied. |Optional

Reviewer: ```suggestion
|Before Calling API |Set to cause your logging policy to log messages before the API endpoint is called, taking into account the order in which this policy is applied |Optional
```

Author: Done! Thanks!

> |After Calling API |Set to cause your logging policy to log messages after the API endpoint is called, taking into account the order in which this policy is applied. |Optional

Reviewer: ```suggestion
|After Calling API |Set to cause your logging policy to log messages after the API endpoint is called, taking into account the order in which this policy is applied |Optional
```

Author: Done! Thanks!
